### ISSUE TITLE
Exceptions thrown in branch muflux, class MufluxReco

### ISSUE THREAD

**Issue by StBies** [ShipSoft/FairShip#316]
# Description of bug
In the class MufluxReco (package charmdet), the method MufluxReco::extrapolateToPlane(genfit::Track* fT,Float_t z, TVector3& pos, TVector3& mom) has a few lines of code where unhandled exceptions can be thrown. For me this happened at least once in RUN_8000_2200, and resulted in a crash.

## Lines that can throw exceptions

https://github.com/ShipSoft/FairShip/blob/126ff278056319178e12f30b084ef1ea991fd605/charmdet/MufluxReco.cxx#L485

https://github.com/ShipSoft/FairShip/blob/126ff278056319178e12f30b084ef1ea991fd605/charmdet/MufluxReco.cxx#L492

https://github.com/ShipSoft/FairShip/blob/126ff278056319178e12f30b084ef1ea991fd605/charmdet/MufluxReco.cxx#L506

https://github.com/ShipSoft/FairShip/blob/126ff278056319178e12f30b084ef1ea991fd605/charmdet/MufluxReco.cxx#L514

## Type of exception?
The exceptions thrown are of type genfit::Exception (https://github.com/ShipSoft/FairShip/blob/muflux/genfit/core/include/Exception.h)

# Where method is called
The method throwing the unhandled exception is called in drifttubeMonitoring.py, line 3094 
https://github.com/ShipSoft/FairShip/blob/126ff278056319178e12f30b084ef1ea991fd605/charmdet/drifttubeMonitoring.py#L3094
 and its result is saved in a variable called tracklength (that is never used from there on).

# Possible way to handle this
In drifttubeMonitoring.py, the same method from c++ is called and there, the exception is handled:
https://github.com/ShipSoft/FairShip/blob/126ff278056319178e12f30b084ef1ea991fd605/charmdet/drifttubeMonitoring.py#L1772

**Comment by ThomasRuf** [comment]
I am also suffering from this problem. But what is the solution? I am not a C++ expert, no idea how to handle exceptions.

**Comment by StBies** [comment]
The way in C++ is to put the code in a block like
```c++
try
{
        auto st  = fT->getFittedState(m);
        auto Pos = st.getPos();
}
catch(const genfit::Exception& e)
{
        //do whatever should happen when getFittedState fails
}


If you can tell me, what should happen, when getFittedState fails I can implement that for you.

Another way is (what automatically happens when exceptions are not handled) that they are passed through to where the throwing function is called. So in this case to python. When python receives an Exception from c++ it automatically converts it to a generic python exception that can also be handled there. This latter option is for example performed in drifttubeMonitoring.py, line 1772

https://github.com/ShipSoft/FairShip/blob/126ff278056319178e12f30b084ef1ea991fd605/charmdet/drifttubeMonitoring.py#L1772

**Comment by olantwin** [comment]
It's probably better to handle the exception in C++, as some information gets lost when python converts it into a generic python exception.

In this case it might be the only one that can be thrown, but more generally it's best to handle the the exception in the most specific way possible (e.g. only handle `genfit::Exception`), to avoid accidentally missing an exception that wasn't expected (e.g. an I/O error) and could indicate other problems.

**Comment by ThomasRuf** [comment]
Could somebody make a pull request? No idea what to do. Should not cost cpu time.

Holen Sie sich Outlook für Android<https://aka.ms/ghei36>




On Thu, Nov 14, 2019 at 10:37 AM +0100, "Oliver Lantwin" <notifications@github.com<mailto:notifications@github.com>> wrote:


It's probably better to handle the exception in C++, as some information gets lost when python converts it into a generic python exception.

In this case it might be the only one that can be thrown, but more generally it's best to handle the the exception in the most specific way possible (e.g. only handle genfit::Exception), to avoid accidentally missing an exception that wasn't expected (e.g. an I/O error) and could indicate other problems.

**Comment by StBies** [comment]
> It's probably better to handle the exception in C++, as some information gets lost when python converts it into a generic python exception.
> 
> In this case it might be the only one that can be thrown, but more generally it's best to handle the the exception in the most specific way possible (e.g. only handle `genfit::Exception`), to avoid accidentally missing an exception that wasn't expected (e.g. an I/O error) and could indicate other problems.

I totally agree.

> Could somebody make a pull request? No idea what to do. Should not cost cpu time.

I will try to figure out what should best happen if an exception occurs and implement it, then pull request. Probably, by simply handling the exception, a lot of the if-checks can be dropped, even saving CPU time.

**Comment by olantwin** [comment]
Handling the exception shouldn't degrade performance. Might be also faster to catch the exception in C++ than waiting for it to propagate to python.

In general exceptions perform similarly to if not better to return-value based error handling, in particular in the non-exceptional case (but there's some debate around that).

**Comment by StBies** [comment]
Is this part of the code:
```c++
if (fstate.getMom()[2] < 900. && fstate.getMom()[2] > 1. && M != 1)
{
        fstate = fT->getFittedState(M);
        pos = fstate.getPos();
	mom = fstate.getMom();
}
else
{
	auto CRep = fT->getCardinalRep();
	auto point = fT->getPointWithMeasurementAndFitterInfo(1,CRep);
	genfit::KalmanFitterInfo* info = dynamic_cast<genfit::KalmanFitterInfo*>(point->getFitterInfo(CRep));
	auto xstate = info->getBackwardUpdate();
	pos = xstate->getPos();
	mom = xstate->getMom();
}
```
really needed or just a workaround of the exception? Especially the condition in the if:
```c++
if (fstate.getMom()[2] < 900. && fstate.getMom()[2] > 1. && M != 1)
```
If not I would just set mom and pos vectors to the backwardUpdate value in case an exception is thrown.

**Comment by StBies** [comment]
# Possible fix, some points should be discussed
I have pushed a possible solution, that seems to give the same results as before.
The code is starting from here:

https://github.com/StBies/FairShip/blob/f76d86178ba2af5858f534cfc3c6e44e0eae3597/charmdet/MufluxReco.cxx#L474

I would appreciate if you cold take a look at the code, especially the TODO comments I introduced. This way I hope, we can minimize the effort it takes to merge a pull request.

**Comment by StBies** [comment]
Seems to be fixed now. Tested on 
RUN_8000_2200, SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root
from the eos data with refit.

Below is a full diff from a log file before and after the changes, residuals did not change.
Should I now do a pull request?
```
-bash-4.2$ diff ../DEBUG_BRANCH_find_thrown_exception9.log ../DEBUG_BRANCH_find_thrown_exception9_catch_fix.log 
7c7                                                                                                                                  
<      initialisation for run id 1573738858                                                                                          
---                                                                                                                                  
>      initialisation for run id 1573804616                                                                                          
57c57                                                                                                                                
< [INFO] Simulation RunID: 1573738858                                                                                                
---                                                                                                                                  
> [INFO] Simulation RunID: 1573804616                                                                                                
72c72                                                                                                                                
< OBJ: TTreeReader      TTreeReader     A simple interface to read trees : 0 at: 0x119da980                                          
---                                                                                                                                  
> OBJ: TTreeReader      TTreeReader     A simple interface to read trees : 0 at: 0x11d268f0                                          
149,156c149,156                                                                                                                      
< now at event 0  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Thu Nov 14 14:46:29 2019                         
< now at event 10000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Thu Nov 14 14:49:14 2019                     
< now at event 20000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Thu Nov 14 14:52:03 2019                     
< now at event 30000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Thu Nov 14 14:54:47 2019                     
< now at event 40000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Thu Nov 14 14:57:28 2019                     
< now at event 50000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Thu Nov 14 15:00:13 2019                     
< now at event 60000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Thu Nov 14 15:02:50 2019                     
< now at event 70000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Thu Nov 14 15:05:27 2019
---
> now at event 0  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Fri Nov 15 08:57:26 2019
> now at event 10000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Fri Nov 15 08:59:39 2019
> now at event 20000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Fri Nov 15 09:01:56 2019
> now at event 30000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Fri Nov 15 09:04:07 2019
> now at event 40000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Fri Nov 15 09:06:17 2019
> now at event 50000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Fri Nov 15 09:08:30 2019
> now at event 60000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Fri Nov 15 09:10:36 2019
> now at event 70000  of  77714 SPILLDATA_8000_0516155320_20180716_081744_RT_refit.root Fri Nov 15 09:12:44 2019
181c181
< timing: {'extrapTrack': 55.80857300758362, 'fillRes': 2.987142324447632, 'prepareTrack': 20.843150854110718, 'analysis': 6.957158803939819, 'fit': 248.56590843200684}
---
> timing: {'extrapTrack': 46.02418303489685, 'fillRes': 2.2713608741760254, 'prepareTrack': 16.639817476272583, 'analysis': 5.355957508087158, 'fit': 184.5211238861084}

```

**Comment by StBies** [comment]
# Request for review
I today tested the code. In fact, it doesn't do exactly the same before as I remove the ```if ( fstate.getMom()[2]<900. && fstate.getMom()[2]>1. && M!=1 )``` check that was labelled with more stable against crashes.
This can result in very slightly different residuals (~ < 1um) but results in improved performance.
Please comment if this is okay or if you find any other errors in the code.

## Code location
The changes can be found here:
https://github.com/StBies/FairShip/blob/5dae9e6333b13d3446b40c93550146e7ae70e9c6/charmdet/MufluxReco.cxx#L474

# Test environment
Tested on RUN 2200 with drifttubeMonitoring.py set with ```zeroField = True``` and running 
```plotBiasedResiduals(PR=13,minP=10.)```
for several spills to check

# Example biased residuals for one Spill 
The files below are for the Spill:
```SPILLDATA_8000_0516150710_20180716_080222_RT_refit.root```
[MUFLUX_3_biased.pdf](https://github.com/ShipSoft/FairShip/files/3851642/MUFLUX_3_biased.pdf)
[CATCH_3_biased.pdf](https://github.com/ShipSoft/FairShip/files/3851646/CATCH_3_biased.pdf)

The plots named MUFLUX are from the unchanged muflux branch, the ones called CATCH from my local branch with the changes. Note that the values differ very slightly.

**Comment by StBies** [comment]
Fixed with #318.
